File: ios/Objective-C/MesiboObjCSample/MesiboObjCSample/MesiboSample.m
//
//  MesiboSample.m
//
//  Copyright © 2020 Mesibo. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "mesibo/mesibo.h"
#import "MesiboSample.h"

@implementation MesiboSample

+(MesiboSample *)getInstance {
    static MesiboSample *myInstance = nil;
    if(nil == myInstance) {
        @synchronized(self) {
            if (nil == myInstance) {
                myInstance = [[self alloc] init];
            }
        }
    }
    return myInstance;
}

-(void) start:(NSString *) accessToken {
    [MesiboInstance setAccessToken:accessToken];
    [MesiboInstance setDatabase:@"mesibo.db" resetTables:0];
    [MesiboInstance addListener:self];
    [MesiboInstance start];
    
}

-(void) sendMessage:(NSString *)to message:(NSString *)message {
    MesiboParams *p = [MesiboParams new];
    
    p.flag = MESIBO_FLAG_DELIVERYRECEIPT | MESIBO_FLAG_READRECEIPT;
    p.expiry = 3600*24*7;
    p.peer = to;
    [MesiboInstance sendMessage:p msgid:[MesiboInstance random] string:message];
}

-(void) readStoredMessagesForUser:(NSString *)address {
    MesiboReadSession *readSession = [MesiboReadSession new];
    [readSession initSession:address groupid:0 query:nil delegate:self];
    [readSession enableSummary:NO];
    [readSession enableReadReceipt:YES];
    [readSession read:100];
}

-(void) readStoredMessagesForGroup:(uint32_t)groupId {
    MesiboReadSession *readSession = [MesiboReadSession new];
    [readSession initSession:nil groupid:groupId query:nil delegate:self];
    [readSession enableSummary:NO];
    [readSession enableReadReceipt:YES];
    [readSession read:100];
}

-(void) Mesibo_OnMessage:(MesiboParams *)params data:(NSData *)data {
    
    
    NSString *message = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

-(void) Mesibo_onFile:(MesiboParams *)params file:(MesiboFileInfo *)file {
    
    
}

-(void) Mesibo_onLocation:(MesiboParams *)params location:(MesiboLocation *)location {
    
}

// so that we get contact while user has started typing
-(void) Mesibo_onActivity:(MesiboParams *)params activity:(int)activity {
}

-(void) Mesibo_OnConnectionStatus:(int)status {
    
    NSLog(@"Connection status: %d", status);
    
    if (MESIBO_STATUS_SIGNOUT == status) {
        
        
    } else if (MESIBO_STATUS_AUTHFAIL == status) {
    }
    
    if(MESIBO_STATUS_ONLINE == status) {
    }
    
}


@end

File: ios/Swift/MesiboSwiftSample/MesiboSwiftSample/MesiboSample.swift
//
//  MesiboSample.swift
//  MesiboSwiftSample
//
//  Copyright © 2020 Mesibo. All rights reserved.
//

import Foundation
import mesibo

@objcMembers public class MesiboSample : NSObject, MesiboDelegate {
    static var getInstanceMyInstance: MesiboSample? = nil
    
    public class func getInstance() -> MesiboSample {
        if nil == getInstanceMyInstance {
            let lockQueue = DispatchQueue(label: "self")
            lockQueue.sync {
                if nil == getInstanceMyInstance {
                    getInstanceMyInstance = MesiboSample()
                    //getInstanceMyInstance?.initialize()
                }
            }
        }
        return getInstanceMyInstance!
    }
    
    func start(_ accessToken: String) {

        Mesibo.getInstance().setAccessToken(accessToken)
        Mesibo.getInstance()!.setDatabase("mesibo.db", resetTables: 0)
        Mesibo.getInstance()!.addListener(self)
        Mesibo.getInstance()!.start();
    }
    
    func sendMessage(_ to:String, message: String) {
        let params: MesiboParams = MesiboParams();
        params.flag = UInt32(MESIBO_FLAG_DELIVERYRECEIPT | MESIBO_FLAG_READRECEIPT);
        params.expiry = 3600*24*7;
        params.peer = to;
        Mesibo.getInstance()?.sendMessage(params, msgid: (Mesibo.getInstance()?.random())!, string: message)
    }
    
    func readStoredMessagesForUsr(_ address: String) {
        let readSession: MesiboReadSession = MesiboReadSession()
        readSession.initSession(address, groupid: 0, query: nil, delegate: self)
        readSession.enableSummary(false)
        readSession.enableReadReceipt(true)
        readSession.read(100)
    }
    
    func readStoredMessagesForGroup(_ groupId: UInt32) {
        let readSession: MesiboReadSession = MesiboReadSession()
        readSession.initSession(nil, groupid: groupId, query: nil, delegate: self)
        readSession.enableSummary(false)
        readSession.enableReadReceipt(true)
        readSession.read(100)
    }
    
    // mesibo delegates
    public func mesibo_(onMessage params: MesiboParams?, data: Data?) {
        
        var message: String? = nil
        if let data = data {
            message = String(data: data, encoding: .utf8)
        }
   
    }
    
    public func mesibo_(onFile params: MesiboParams?, file: MesiboFileInfo?) {
        
    }
    
    public func mesibo_(onLocation params: MesiboParams?, location: MesiboLocation?) {
        
    }
    
    public func mesibo_(onActivity params: MesiboParams?, activity: Int32) {
    }
    
    public func mesibo_(onConnectionStatus status: Int32) {
        
        if MESIBO_STATUS_SIGNOUT == status {
            
        } else if MESIBO_STATUS_AUTHFAIL == status {
        }
        
        if MESIBO_STATUS_ONLINE == status {
            print("You are online")
        }
    }

}
